### PR TITLE
CLI

### PR DESCRIPTION
**Pull request by novalis**
From my perspective, there is still some to to do here, but I think it might be worth merging, after code review:

- [x] I need to understand why the michelson stuff is using Reactive and maybe fix it

- [x] There are a ton of unused variables which need to be replaced

- [x] Error handling is kind of janky and often badly done

- [x] Speaking of error handling, we need to print warnings if we get any

- [x] JSON output

- [x]  raw output (including figuring out what "raw" means)

- [x] Understand the big_map stuff

- [x] Figure out if we really need everything in Import to be there.

- [x] I think I want to improve the logging further -- right now, we have ctxt#formatter, but I think we might want to accrete prefixes, instead of passing logging functions around.

- [x] More tests -- right now, I promoted a few existing tests, but in order to promote the rest, I need to submit a PR to base58 and get it accepted.
  a. Speaking of which, should we consider doing something with the base58 module on opam?  Like, contacting the authors to accept our patches?

- [x] Squash history -- of course, this is a mess because I was adding and removing a bunch of fixmes, and because I didn't have formatting set up, and because I was making global changes.  No need for all of that.

### CODE REVIEW

**Comment by novalis** [comment]
Commented on https://github.com/mirage/ezjsonm/pull/45 
Filed https://github.com/vbmithr/ocaml-base58/pull/1

File: src/lib/contract_metadata.ml [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+          | Failure f -> error `Network s f
+          | e -> Fmt.kstr (error `Network s) "%a" Exn.pp e ) ;
+          Ok () in
+    let uri =
+      Uri.of_string uri_code |> of_uri ~validate_kt1_address ~validate_network
+    in
+    (uri, List.rev !errors)
+
+  let%test "validate_address bogus-schema" =
+    (* Here, we expect  validate_address to return Ok even though it has some errors.  FIXME: why?*)
```
Comment (novalis) [inline src/lib/contract_metadata.ml]: Why does this test pass?  I don't mean from a code perspective -- I mean why does a function named validate_address return success here?  Should we fix this?

Comment (smondet) [reply src/lib/contract_metadata.ml]: It does indeed deserve a better name and/or a better structured return value. But in that case it should return `Error _, _`; a wrong scheme is a hard failure.

See
https://github.com/oxheadalpha/tezos-contract-metadata/blob/4c1f5dc0b2d27d1c096c78bf49ea36e0975f6492/src/lib/metadata_uri.ml#L196

File: src/cli/main.ml [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+  let format =
+    Arg.enum
+      [ ("text", Text Full)
+      ; ("text:full", Text Full)
+      ; ("text:short", Text Short)
+      ; ("raw", Raw)
+      ; ("json", Json) ] in
+  Arg.(
+    value
+    & opt format ~vopt:(Text Full) (Text Full)
```
Comment (smondet) [inline src/cli/main.ml]: Not sure the use of `~vopt` fits well there. It means that passing just `--format` without an argument is equivalent to not passing it at all.

File: src/cli/main.ml [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+(* CLI *)
+let metadata_format =
+  let doc = "metadata format" in
+  let format =
+    Arg.enum
+      [ ("text", Text Full)
+      ; ("text:full", Text Full)
+      ; ("text:short", Text Short)
+      ; ("raw", Raw)
+      ; ("json", Json) ] in
```
Comment (smondet) [inline src/cli/main.ml]: this is redundant with the function `string_of_output_format` above, better have them in one place.

File: src/cli/main.ml [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    & opt format ~vopt:(Text Full) (Text Full)
+    & info ["format"] ~docv:"FORMAT" ~doc)
+
+let debug =
+  let doc = "Debugging output to stderr." in
+  let yes = (true, Arg.info ["debug"] ~doc) in
+  Arg.(last & vflag_all [false] [yes])
+
+let src =
+  let doc = "source" in
```
Comment (smondet) [inline src/cli/main.ml]: We'd need something more descriptive here.

In this case the source is a KT1 contract on one of the networks we support.

File: src/cli/main.ml [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+let show_metadata src format debug =
+  let ctxt =
+    let nodes = Query_nodes.Node_list.nodes (Query_nodes.get_default_nodes ()) in
+    let now () = Unix.gettimeofday () /. 1000. in
+    let time_zero = now () in
+    object (self)
+      val prefix = ""
+      method nodes = nodes
+      method formatter = if debug then Fmt.stderr else Caml.Format.str_formatter
```
Comment (smondet) [inline src/cli/main.ml]: For a formatter that does nothing it's better to use `Caml.Format.make_formatter (fun _ _ _ -> ()) (fun () -> ())` (the `str_formatter` is going to be an internal buffer that keeps on growing).

File: src/cli/main.ml [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+      method sleep = Lwt_unix.sleep
+      method http_timeout () = 5.0
```
Comment (smondet) [inline src/cli/main.ml]: Also tricky annoying ocaml thing:
→ object methods are lazy, the `()` is confusing/redundant there.
And the `if debug` above is evaluated every time somebody uses `c#formatter`.

File: src/lib/tezos-contract-metadata.opam [hunk 7]
```diff
@@ -6,12 +6,23 @@ bug-reports: "https://github.com/oxheadalpha/tezos-contract-metadata/issues"
 dev-repo: "git+https://github.com/oxheadalpha/tezos-contract-metadata.git"
 license: "MIT"
 depends: [
+  "base"
+  "base58"
+  "digestif" { = "0.9.0" }
+  "lwt"
+  "lwt_ssl"
```
Comment (smondet) [inline src/lib/tezos-contract-metadata.opam]: I think there is a way to depend either on `lwt_ssl` or on `tls`, but if we pick one I think `tls` is safer.

File: src/lib/tezos-contract-metadata.opam [hunk 8]
```diff
@@ -6,12 +6,23 @@ bug-reports: "https://github.com/oxheadalpha/tezos-contract-metadata/issues"
...
+  "ppx_inline_test"
+  "ppx_expect"
+  "cohttp-lwt-unix"
```
Comment (smondet) [inline src/lib/tezos-contract-metadata.opam:16]: Maybe not in this PR, but the cli app should have its own opam file that depends on unix stuff and on the library.

File: src/cli/main.ml (hunk 3 above)
Comment (novalis) [reply src/cli/main.ml]: This is the opposite direction.  That is string -> format, and this is format -> string.  Is there a good way to have that bidirectional association.

File: src/lib/tezos-contract-metadata.opam (hunk 8 above)
Comment (novalis) [reply src/lib/tezos-contract-metadata.opam:16]: I understand why you want to separate those deps out -- it makes sense that the lib should not depend on lwt_unix since we want it to be buildable with jsoo.  

I want to understand the way that we organize our code and releases.  The units of organization I am concerned about are (a) opam packages and (b) repositories.   Is it usual for one repository to contain multiple opam packages?  And how does one describe a dependency (from the CLI) on the HEAD version (that is, the version that corresponds to the CLI's version) of the lib?  Or is the idea that developers have to manually set up that dependency using some manual opam stuff like opam monorepo, and then remember to sequence releases correctly?  I'm used to trunk-based development, where src/cli's HEAD version depends on src/lib's HEAD version (which, in turn, would depend on the HEAD versions of all of their deps).  But I know that's not the norm outside of big corps.

Basically, I want it to be easy for someone coming upon our code to submit a patch that might cross the lib/cli boundary, and to keep any invariants about versioning in place.

File: src/cli/main.ml (hunk 3 above)
Comment (smondet) [reply src/cli/main.ml]: one way can be just using:

```ocaml
List.map [Text Full; Text Short; Raw; Json]
  ~f:(fun v -> string_of_output_format v, v)
```

or the other way around: “make `string_of_output_format` search in a list” but that a bit annoying because of the error case and the fact that we don't get help form the exhaustiveness check

In the case were performance is really important (I'd say not here :) ), at least centralizing both cases in a small module helps with maintnance.

File: src/lib/tezos-contract-metadata.opam (hunk 8 above)
Comment (smondet) [reply src/lib/tezos-contract-metadata.opam:16]: yes many opam packages within a same repo is OK

one thing that helps keeping things in sync is getting `dune` to generate the opam files, cf. https://dune.readthedocs.io/en/stable/opam.html#generating-opam-files

not being a "big-corp" we need to keep balancing between being monorepo and package-manager -friendly; not sure we have it all figured out but at least `dune` is designed that way.

Comment (novalis) [reply src/lib/tezos-contract-metadata.opam:16]: While dune knows how to generate opam files, I am not sure that it knows how to update their versions as needed.  But maybe the release-to-opam process could do that.  I'll play with it at some point, when we're closer to a release.

File: src/lib/contract_metadata.ml [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        let j = Ezjsonm.value_from_string s in
+        let rec fix = function
+          | (`String _ | `Float _ | `Bool _ | `Null) as v -> v
+          | `A l -> `A (List.map l ~f:fix)
+          | `O kvl ->
+              let f (k, v) =
+                let fix_warn o k =
+                  ( match
+                      List.exists !warnings ~f:(function
+                          | `Fixed_legacy (a, _) -> String.equal a o )
```
Comment (novalis) [inline src/lib/contract_metadata.ml]: What does "fixed_legacy' mean here?

Comment (smondet) [reply src/lib/contract_metadata.ml]: cf. just below:

```ocaml
match k with
| "michelson-storage-view" -> fix_warn k "michelsonStorageView"
| "return-type" -> fix_warn k "returnType"
| other -> (other, fix v) in
```

early draft versions of TZIP-16 were using kebab-case and at some point it was decided to switch to CamelCase, so I added this little hack to TZComet to change errors into warnings for the contracts that were already being originated/tested.

I think we could remove that now, we'd have to double check, but so far it hasn't hurt either (other that readability of the code ;) ) so it's still around

File: README.md [hunk 10]
```diff
@@ -27,6 +27,27 @@ Linting:
 dune build @fmt --auto-promote
 ```
 
+## Build (cli)
+
+opam install src/lib/tezos-contract-metadata.opam # Install the lib to
+                                                  # provide the dep for the CLI
```
Comment (smondet) [inline README.md]: you shouldn't need this to build, dune should pick up the dependency from the tree. Same for the tests below.

File: src/cli/dune-project [hunk 11]
```diff
@@ -0,0 +1,28 @@
+(lang dune 2.9)
+(name tezos-contract-metadata-cli)
+
+(generate_opam_files true)
+
+(source (github oxheadalpha/tezos-contract-metadata))
+(license MIT)
+(authors "Tezos Devteam")
+(maintainers "david.turner@oxheadalpha.com")
+
```
Comment (smondet) [inline src/cli/dune-project]: shouldn't all of this be in the toplevel `dune-project` in order to share between cli and lib?

File: src/lib/contract_metadata.ml [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+                  add_if_not_present ~list:warnings
+                    ~present:(function
+                      | `Fixed_legacy (a, _) -> String.equal a key )
+                    ~new_value:(`Fixed_legacy (key, location)) ;
+                  (location, fix v) in
+                match k with
+                | "michelson-storage-view" ->
+                    fix_warn k " should be spelled \"michelsonStorageView\""
+                | "return-type" ->
+                    fix_warn k " should be spelled \"returnType\""
```
Comment (smondet) [inline src/lib/contract_metadata.ml]: this will make the parser fail with a weird error message now?

- the second argument is not _(just)_ a location, it's the replacement json-object key to feed to `Json_encoding.destruct`

File: src/lib/import.ml [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+  type 'a t = 'a
+    constraint
+      'a =
+      < formatter: Caml.Format.formatter
+      ; log_context: string
+      ; http_client: Http_client.t
+      ; program_time: float
+      ; with_log_context: string -> 'a t
+      ; nodes: Query_node.Node.t list
+      ; .. >
```
Comment (smondet) [inline src/lib/import.ml:95]: Not necessary to fix now but this makes the dependency tracking _less_ precise, e.g. every signature that takes a `_ Context.t` now claims to require the `http_client` even if it does not use it in practice.

File: src/lib/metadata_contents.ml [hunk 14]
```diff
@@ -798,3 +798,222 @@ laborum.
       with _ -> [] in
     o 0
 end
+
+let%test "empty metadata json" =
+  let examples = Example.all () in
+  let example = List.nth examples 0 in
+  String.equal (to_json example) "{}"
+
+let%test "short metadata json" =
```
Comment (smondet) [inline src/lib/metadata_contents.ml]: again, no need to fix this time, but these kinds of "big blob" tests fit better the `let%expect_test` model (and dune provides automation to update them on-demand).

File: src/lib/metadata_uri.ml [hunk 15]
```diff
@@ -159,13 +164,24 @@ let rec of_uri ?validate_network ?validate_kt1_address uri =
...
       >>? fun (network, address) ->
-      let validate_option f v =
-        match (f, v) with
-        | Some (validate : field_validation), Some value -> validate value
-        | _, _ -> ok () in
-      validate_option validate_network network
+      let validate_option ~f v =
+        match v with Some value -> f value | None -> ok () in
+      validate_option network ~f:(function
+        | "mainnet" | "carthagenet" | "delphinet" | "dalphanet" | "zeronet" ->
```
Comment (smondet) [inline src/lib/metadata_uri.ml]: at some point we need a way to keep this list properly up to date

right now `"edonet"`, `"florencenet"`, `"granadanet"`, `"hangzhounet"`, `"mondaynet"` are relevant values.

but that's another reason to not have it here, see next comment :arrow_down:

File: src/lib/metadata_uri.ml [hunk 16]
```diff
@@ -159,13 +164,24 @@ let rec of_uri ?validate_network ?validate_kt1_address uri =
...
+              Fmt.kstr
+                (fun x -> fail (Wrong_network (network, x)))
+                "%a" Base.Exn.pp e ) )
       >>? fun () ->
-      validate_option validate_kt1_address address
+      validate_option address ~f:(fun address ->
+          try Ok (ignore (B58_hashes.check_b58_kt1_hash address)) with
+          | Failure f -> fail (Bad_b58 (address, f))
+          | e ->
+              Fmt.kstr (fun x -> fail (Bad_b58 (address, x))) "%a" Base.Exn.pp e )
```
Comment (smondet) [inline src/lib/metadata_uri.ml]: Why inline these two? the point was not to impose base58/hash functions on the user side.

File: src/lib/michelson_bytes.ml [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+    ("SLICE", "SLICE")
+  ; (* Alpha_005 addition *)
+    ("DIG", "DIG")
+  ; ("DUG", "DUG")
+  ; ("EMPTY_BIG_MAP", "EMPTY_BIG_MAP")
+  ; ("APPLY", "APPLY")
+  ; ("chain_id", "chain_id")
+  ; ("CHAIN_ID", "CHAIN_ID")
+    (* New instructions must be added here, "for" backward compatibility of the encoding. *)
+  ]
```
Comment (smondet) [inline src/lib/michelson_bytes.ml:163]: another list we'll have to update at some point

File: README.md (hunk 10 above)
Comment (novalis) [reply README.md]: Hm, weird -- this wasn't working before, but it is now.  OK, will fix.

File: src/lib/contract_metadata.ml (hunk 12 above)
Comment (novalis) [reply src/lib/contract_metadata.ml]: oh, I see

File: src/lib/import.ml (hunk 13 above)
Comment (novalis) [reply src/lib/import.ml:95]: Basically all of the signatures require a `http_client` anyway.  And we need something like this to define `with_log_context` (else what would its return-type be?).  Context is sort of a weird object here -- most of it is "dependency injection", and then only the log context is genuinely contextual.  It looks like some people recommend functors for dependency injection, but I think there would still either be giant signatures or some bundling.

File: src/lib/metadata_uri.ml (hunk 16 above)
Comment (novalis) [reply src/lib/metadata_uri.ml]: I don't understand why you wouldn't want these checks.  Why would you ever want different checks?

Comment (smondet) [reply src/lib/metadata_uri.ml]: you may want to use different libraries there (pure ocaml or javascript bindings with JSOO, or the native tezos-crypto library if you have it in scope, etc)

the question was also: since the effort of parametrizing them was already done, is there a gain in going backwards?

File: src/lib/michelson_bytes.ml (hunk 17 above)
Comment (novalis) [reply src/lib/michelson_bytes.ml:163]: Cool, is there a lib we could get these from, rather than maintaining our own?  I see [tezos-protocol-alpha](https://opam.ocaml.org/packages/tezos-protocol-alpha/), but it seems to depend on the entire tezos universe, which seems like a shame.  Pulling out just michelson_v1_primitives.ml would almost work, but it depends on the error monad, and I think that's going to drag in a ton of stuff. Instead, michelson_v1_primitives.ml would have to pull in a separate lib that's just the prim<->string mapping.  This is definitely something to defer.

File: src/lib/metadata_uri.ml (hunk 16 above)
Comment (novalis) [reply src/lib/metadata_uri.ml]: Oh, I see -- you don't want to change the functionality, just the implementation.  So it's another sort of dependency injection.  I didn't notice that base58 depended on zarith (which depends on gmp).  Normally, I would not expect a core algorithm to be injected.  

I think the reason I went to the non-parameterized was to improve the error handling and simplify the code.  Even if we do want it injected, I think the thing to do is to inject the base58 module instead?

File: src/lib/michelson_bytes.ml (hunk 17 above)
Comment (smondet) [reply src/lib/michelson_bytes.ml:163]: just so you know, NL has been steadily working on being able to build the protocol with a valid environment (= the hard part) under JSOO, this list is suuposed to be a temporary hack while waiting (it is also nice not to depend on a specific version of the protocol though)

File: src/lib/metadata_uri.ml (hunk 16 above)
Comment (smondet) [reply src/lib/metadata_uri.ml]: well the implementation is most obvious (and the one that matters in practice) but those two checks can be understood differently regarding functionality too, depending on context: is a KT1 just a valid b58hash or a KT1 that actuallly exists within a given network? the list of available networks has a meaning for public networks, but some private-chain deployment could have their own set of names, ... passing the functions as argument was a way of not having to make a decision in the central library.

Comment (novalis) [reply src/lib/metadata_uri.ml]: Out of curiosity, I tried rewriting base58 to not use zarith, and for short strings the perf was the same, but for long strings the new version was significantly worse (after my best quick shot at micro-optimization but no algorithmic cleverness). I guess gmp is pretty optimized. Maybe it doesn't matter because it's only used for short strings, but that still seems like a patch I would be unlikely to get merged. I'll put it back the way it was.

File: src/lib/tezos-contract-metadata.opam (hunk 8 above)
Comment (novalis) [reply src/lib/tezos-contract-metadata.opam:16]: OK, I did this.

File: src/lib/contract_metadata.ml (hunk 12 above)
Comment (novalis) [reply src/lib/contract_metadata.ml]: fixed.

File: src/lib/metadata_contents.ml (hunk 14 above)
Comment (novalis) [reply src/lib/metadata_contents.ml]: done

File: src/lib/metadata_uri.ml (hunk 16 above)
Comment (novalis) [reply src/lib/metadata_uri.ml]: restored
